vm/ubpf_vm.c

```diff
@@ -1504,6 +1583,50 @@ validate(const struct ubpf_vm* vm, const struct ebpf_inst* insts, uint32_t num_i
         case EBPF_OP_MOD64_IMM:
             break;
 
+        // 64-bit atomic operations
+        case EBPF_OP_ATOMIC_STORE: {
+            store = true;
+            switch (inst.imm & EBPF_ALU_OP_MASK) {
+            case EBPF_ALU_OP_ADD:
+                break;
+            case EBPF_ALU_OP_OR:
+                break;
+            case EBPF_ALU_OP_AND:
+                break;
+            case EBPF_ALU_OP_XOR:
+                break;
+            case (EBPF_ATOMIC_OP_XCHG & ~EBPF_ATOMIC_OP_FETCH):
+                break;
+            case (EBPF_ATOMIC_OP_CMPXCHG & ~EBPF_ATOMIC_OP_FETCH):
+                break;
+            default:
+                *errmsg = ubpf_error("invalid atomic operation at PC %d", i);
+                return false;
+            }
+            break;
+        }
+        // 32-bit atomic operations
+        case EBPF_OP_ATOMIC32_STORE: {
+            store = true;
+            switch (inst.imm & EBPF_ALU_OP_MASK) {
+            case EBPF_ALU_OP_ADD:
+                break;
+            case EBPF_ALU_OP_OR:
+                break;
+            case EBPF_ALU_OP_AND:
+                break;
+            case EBPF_ALU_OP_XOR:
+                break;
+            case (EBPF_ATOMIC_OP_XCHG & ~EBPF_ATOMIC_OP_FETCH):
+                break;
+            case (EBPF_ATOMIC_OP_CMPXCHG  & ~EBPF_ATOMIC_OP_FETCH):
+                break;
+            default:
+                *errmsg = ubpf_error("invalid atomic operation at PC %d", i);
```

Comment:
In a similar error message above, we print out the opcode that is invalid. Do we want to do that here, too?